// smart-webcomponents-angular/esm5/form/smart.form.js
import * as tslib_1 from "tslib";
import { Component, Directive, AfterViewInit, ElementRef, Input, OnInit, OnChanges, OnDestroy, SimpleChanges, Output, EventEmitter, QueryList, ContentChildren } from '@angular/core';
import { BaseElement, Smart } from './smart.element';
export { Smart } from './smart.element';
var FormComponent = /** @class */ (function (_super) {
    tslib_1.__extends(FormComponent, _super);
    function FormComponent(ref) {
        var _this = _super.call(this, ref) || this;
        _this.eventHandlers = [];
        _this.nativeElement = ref.nativeElement;
        return _this;
    }
    /** @description Creates the component on demand.
     * @param properties An optional object of properties, which will be added to the template binded ones.
     */
    FormComponent.prototype.createComponent = function (properties) {
        if (properties === void 0) { properties = {}; }
        this.nativeElement = document.createElement('smart-form');
        for (var propertyName in properties) {
            this.nativeElement[propertyName] = properties[propertyName];
        }
        return this.nativeElement;
    };
    Object.defineProperty(FormComponent.prototype, "columns", {
        /** @description Sets or gets the form columns. */
        get: function () {
            return this.nativeElement ? this.nativeElement.columns : undefined;
        },
        set: function (value) {
            this.nativeElement ? this.nativeElement.columns = value : undefined;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FormComponent.prototype, "controls", {
        /** @description Sets or gets the form controls. */
        get: function () {
            return this.nativeElement ? this.nativeElement.controls : undefined;
        },
        set: function (value) {
            this.nativeElement ? this.nativeElement.controls = value : undefined;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FormComponent.prototype, "onStatusChanges", {
        /** @description Callback function for handling status changes. The status could be  'disabled', 'invalid', 'pending', 'valid'. */
        get: function () {
            return this.nativeElement ? this.nativeElement.onStatusChanges : undefined;
        },
        set: function (value) {
            this.nativeElement ? this.nativeElement.onStatusChanges = value : undefined;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FormComponent.prototype, "onValueChanges", {
        /** @description Callback function for handling value changes */
        get: function () {
            return this.nativeElement ? this.nativeElement.onValueChanges : undefined;
        },
        set: function (value) {
            this.nativeElement ? this.nativeElement.onValueChanges = value : undefined;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FormComponent.prototype, "labelPosition", {
        /** @description Sets or Gets the labels position. */
        get: function () {
            return this.nativeElement ? this.nativeElement.labelPosition : undefined;
        },
        set: function (value) {
            this.nativeElement ? this.nativeElement.labelPosition = value : undefined;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FormComponent.prototype, "readonly", {
        /** @description Makes the form readonly. */
        get: function () {
            return this.nativeElement ? this.nativeElement.readonly : undefined;
        },
        set: function (value) {
            this.nativeElement ? this.nativeElement.readonly = value : undefined;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FormComponent.prototype, "showColonAfterLabel", {
        /** @description Shows / hides the colon after the labels. */
        get: function () {
            return this.nativeElement ? this.nativeElement.showColonAfterLabel : undefined;
        },
        set: function (value) {
            this.nativeElement ? this.nativeElement.showColonAfterLabel = value : undefined;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FormComponent.prototype, "showSummary", {
        /** @description Shows / hides validation summary. */
        get: function () {
            return this.nativeElement ? this.nativeElement.showSummary : undefined;
        },
        set: function (value) {
            this.nativeElement ? this.nativeElement.showSummary = value : undefined;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FormComponent.prototype, "state", {
        /** @description Gets the Form's state. Each member in the state has { dirty, untouched, disabled } properties. */
        get: function () {
            return this.nativeElement ? this.nativeElement.state : undefined;
        },
        set: function (value) {
            this.nativeElement ? this.nativeElement.state = value : undefined;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FormComponent.prototype, "value", {
        /** @description Gets or Sets the Form value. */
        get: function () {
            return this.nativeElement ? this.nativeElement.value : undefined;
        },
        set: function (value) {
            this.nativeElement ? this.nativeElement.value = value : undefined;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FormComponent.prototype, "validateOnLoad", {
        /** @description Automatically validates the form when it is created. */
        get: function () {
            return this.nativeElement ? this.nativeElement.validateOnLoad : undefined;
        },
        set: function (value) {
            this.nativeElement ? this.nativeElement.validateOnLoad = value : undefined;
        },
        enumerable: true,
        configurable: true
    });
    /** @description Adds a control to the Form.
    * @param {any} controlOptions. Control options. The control options description is available in the <em>controls</em> property.
    */
    FormComponent.prototype.addControl = function (controlOptions) {
        var _this = this;
        if (this.nativeElement.isRendered) {
            this.nativeElement.addControl(controlOptions);
        }
        else {
            this.nativeElement.whenRendered(function () {
                _this.nativeElement.addControl(controlOptions);
            });
        }
    };
    /** @description Gets a control by its name(dataField).
    * @param {string} dataField. dataField of a FormControl or FormGroup
    * @returns {Control}
  */
    FormComponent.prototype.getControl = function (dataField) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var getResultOnRender, result;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        getResultOnRender = function () {
                            return new Promise(function (resolve) {
                                _this.nativeElement.whenRendered(function () {
                                    var result = _this.nativeElement.getControl(dataField);
                                    resolve(result);
                                });
                            });
                        };
                        return [4 /*yield*/, getResultOnRender()];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, result];
                }
            });
        });
    };
    /** @description Inserts a control to the Form.
    * @param {number} index. Control insert index
    * @param {any} controlOptions. Control options. The control options description is available in the <em>controls</em> property.
    */
    FormComponent.prototype.insertControl = function (index, controlOptions) {
        var _this = this;
        if (this.nativeElement.isRendered) {
            this.nativeElement.insertControl(index, controlOptions);
        }
        else {
            this.nativeElement.whenRendered(function () {
                _this.nativeElement.insertControl(index, controlOptions);
            });
        }
    };
    /** @description Remove a control from the Form.
    * @param {any} controlOptions. Control options. The control options description is available in the <em>controls</em> property.
    */
    FormComponent.prototype.removeControl = function (controlOptions) {
        var _this = this;
        if (this.nativeElement.isRendered) {
            this.nativeElement.removeControl(controlOptions);
        }
        else {
            this.nativeElement.whenRendered(function () {
                _this.nativeElement.removeControl(controlOptions);
            });
        }
    };
    /** @description Submits the form.
    * @param {any} submitOptions?. Sets the submit options object. The object may have the following properties: <em>async</em>, <em>action</em>, <em>target</em>, <em>method</em>. <em>async</em> determines whether the form will be submitted asynchronously. <em>action</em> determines the submit url, <em>method</em> sets whether the submit is through 'GET' or 'POST'. <em>target</em> determines the submit target.
    */
    FormComponent.prototype.submit = function (submitOptions) {
        var _this = this;
        if (this.nativeElement.isRendered) {
            this.nativeElement.submit(submitOptions);
        }
        else {
            this.nativeElement.whenRendered(function () {
                _this.nativeElement.submit(submitOptions);
            });
        }
    };
    /** @description Clears the form.
    */
    FormComponent.prototype.reset = function () {
        var _this = this;
        if (this.nativeElement.isRendered) {
            this.nativeElement.reset();
        }
        else {
            this.nativeElement.whenRendered(function () {
                _this.nativeElement.reset();
            });
        }
    };
    /** @description Validates the form.
    */
    FormComponent.prototype.validate = function () {
        var _this = this;
        if (this.nativeElement.isRendered) {
            this.nativeElement.validate();
        }
        else {
            this.nativeElement.whenRendered(function () {
                _this.nativeElement.validate();
            });
        }
    };
    Object.defineProperty(FormComponent.prototype, "isRendered", {
        get: function () {
            return this.nativeElement ? this.nativeElement.isRendered : false;
        },
        enumerable: true,
        configurable: true
    });
    FormComponent.prototype.ngOnInit = function () {
    };
    FormComponent.prototype.ngAfterViewInit = function () {
        var that = this;
        that.onCreate.emit(that.nativeElement);
        Smart.Render();
        this.nativeElement.classList.add('smart-angular');
        this.nativeElement.whenRendered(function () { that.onReady.emit(that.nativeElement); });
        this.listen();
    };
    FormComponent.prototype.ngOnDestroy = function () {
        this.unlisten();
    };
    FormComponent.prototype.ngOnChanges = function (changes) {
        if (this.nativeElement && this.nativeElement.isRendered) {
            for (var propName in changes) {
                if (changes.hasOwnProperty(propName)) {
                    this.nativeElement[propName] = changes[propName].currentValue;
                }
            }
        }
    };
    /** @description Add event listeners. */
    FormComponent.prototype.listen = function () {
        var that = this;
    };
    /** @description Remove event listeners. */
    FormComponent.prototype.unlisten = function () {
        var that = this;
    };
    FormComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    tslib_1.__decorate([
        Input()
    ], FormComponent.prototype, "columns", null);
    tslib_1.__decorate([
        Input()
    ], FormComponent.prototype, "controls", null);
    tslib_1.__decorate([
        Input()
    ], FormComponent.prototype, "onStatusChanges", null);
    tslib_1.__decorate([
        Input()
    ], FormComponent.prototype, "onValueChanges", null);
    tslib_1.__decorate([
        Input()
    ], FormComponent.prototype, "labelPosition", null);
    tslib_1.__decorate([
        Input()
    ], FormComponent.prototype, "readonly", null);
    tslib_1.__decorate([
        Input()
    ], FormComponent.prototype, "showColonAfterLabel", null);
    tslib_1.__decorate([
        Input()
    ], FormComponent.prototype, "showSummary", null);
    tslib_1.__decorate([
        Input()
    ], FormComponent.prototype, "state", null);
    tslib_1.__decorate([
        Input()
    ], FormComponent.prototype, "value", null);
    tslib_1.__decorate([
        Input()
    ], FormComponent.prototype, "validateOnLoad", null);
    FormComponent = tslib_1.__decorate([
        Directive({
            exportAs: 'smart-form', selector: 'smart-form, [smart-form]'
        })
    ], FormComponent);
    return FormComponent;
}(BaseElement));
export { FormComponent };
